lib/Relation.php

```diff
@@ -0,0 +1,557 @@
+     *
+     * @throws RecordNotFound if any of the records cannot be found
+     *
+     * @return Model|array<Model>
+     */
+    public function find()
+    {
+        $args = func_get_args();
+        $num_args = count($args);
+        $class = $this->className;
+
+        if (0 === $num_args) {
+            throw new ValidationsArgumentError('find requires at least one argument');
+        }
+
+        // find by pk
+        if (1 === $num_args) {
+            $args = $args[0];
+        }
+
+        $single = !is_array($args) || !array_is_list($args);
+
+        $options = $this->options;
+        $options['conditions'] ??= [];
+        $options['conditions'][] = $this->pk_conditions($args);
+        $options['mapped_names'] = $this->alias_attribute;
+
+        $list = $this->table()->find($options);
+        if (is_array($args) && count($list) != count($args)) {
+            throw new RecordNotFound('found ' . count($list) . ', but was looking for ' . count($args));
+        }
+
+        return $single ? ($list[0] ?? throw new RecordNotFound('tbd')) : $list;
+    }
+
+    /**
+     * Returns a record (or N records if a parameter is supplied) without any implied
+     * order. The order will depend on the database implementation.
+     * If an order is supplied it will be respected.
+     *
+     *   Person::take() # returns an object fetched by SELECT * FROM people LIMIT 1
+     *   Person::take(5) # returns 5 objects fetched by SELECT * FROM people LIMIT 5
+     *   Person::where(["name LIKE '%?'", name])->take()
+     */
+    public function take(int $limit = null): mixed
+    {
+        $this->limit($limit ?? 1);
+        if (!isset($limit)) {
+            $models = $this->to_a();
+
+            return $models[0] ?? null;
+        }
+
+        return $this->to_a();
+    }
+
+    /**
+     * Find the first record (or first N records if a parameter is supplied).
+     * If no order is defined it will order by primary key.
+     *
+     * Person::first() // returns the first object fetched by SELECT * FROM people ORDER BY people.id LIMIT 1
+     * Person::where(["user_name = ?", user_name])->first()
+     * Person::where(["user_name = :u", { u: user_name }])->first()
+     * Person::order("created_on DESC")->offset(5).first()
+     * Person.first(3) // returns the first three objects fetched by SELECT * FROM people ORDER BY people.id LIMIT 3
+     */
+    public function first(int $limit = null): mixed
+    {
+        $this->limit($limit ?? 1);
+
+        $pk = $this->table()->pk;
+        if (!empty($pk) && !array_key_exists('order', $this->options) && !array_key_exists('group', $this->options)) {
+            $this->options['order'] = implode(' ASC, ', $this->table()->pk) . ' ASC';
+        }
+
+        if (!isset($limit)) {
+            $models = $this->to_a();
+
+            return $models[0] ?? null;
+        }
+
+        return $this->to_a();
```

Comment:
Something like last()'s

return isset($limit) ? $models : $models[0] ?? null;

would be better

Reply:
Sure, why not? 6a5a0db